kernel/tests/golden_tables.rs

```diff
@@ -0,0 +1,343 @@
+//! Golden table tests adapted from delta-io/delta tests.
+//!
+//! Data (golden tables) are stored in tests/golden_data/<table_name>.tar.zst
+//! Each table directory has a table/ and expected/ subdirectory with the input/output respectively
+
+use arrow::{compute::filter_record_batch, record_batch::RecordBatch};
+use arrow_array::ArrayRef;
+use arrow_select::concat::concat_batches;
+use paste::paste;
+use std::path::Path;
+use std::sync::Arc;
+
+use arrow_schema::SchemaRef as ArrowSchemaRef;
+use delta_kernel::{engine::arrow_data::ArrowEngineData, DeltaResult, Table};
+use futures::{stream::TryStreamExt, StreamExt};
+use object_store::{local::LocalFileSystem, ObjectStore};
+use parquet::arrow::async_reader::{ParquetObjectReader, ParquetRecordBatchStreamBuilder};
+
+use arrow_array::Array;
+use arrow_schema::DataType;
+use delta_kernel::engine::default::executor::tokio::TokioBackgroundExecutor;
+use delta_kernel::engine::default::DefaultEngine;
+
+#[macro_use]
+mod common;
+use common::to_arrow;
+
+// load the test data from <test_table>.tar.zst file
+fn load_test_data(test_name: &str) -> Result<tempfile::TempDir, Box<dyn std::error::Error>> {
+    let path = format!("tests/golden_data/{}.tar.zst", test_name);
+    let tar = zstd::Decoder::new(std::fs::File::open(path)?)?;
+    let mut archive = tar::Archive::new(tar);
+    let temp_dir = tempfile::tempdir()?;
+    archive.unpack(temp_dir.path())?;
+    Ok(temp_dir)
+}
+
+// NB adapated from DAT: read all parquet files in the directory and concatenate them
+async fn read_expected(path: &Path) -> DeltaResult<Option<RecordBatch>> {
+    let store = Arc::new(LocalFileSystem::new_with_prefix(path)?);
+    let files = store.list(None).try_collect::<Vec<_>>().await?;
+    let mut batches = vec![];
+    let mut schema = None;
+    for meta in files.into_iter() {
+        if let Some(ext) = meta.location.extension() {
+            if ext == "parquet" {
+                let reader = ParquetObjectReader::new(store.clone(), meta);
+                let builder = ParquetRecordBatchStreamBuilder::new(reader).await?;
+                if schema.is_none() {
+                    schema = Some(builder.schema().clone());
+                }
+                let mut stream = builder.build()?;
+                while let Some(batch) = stream.next().await {
+                    batches.push(batch?);
+                }
+            }
+        }
+    }
+    let all_data = concat_batches(&schema.unwrap(), &batches)?;
+    Ok(Some(all_data))
+}
+
+// TODO: change to do something similar to dat tests instead of string comparison
```

Comment:
nit: explain what dat does. i.e. "change to use arrow's column `Eq` like dat does, instead of string comparison. Should print out string when test fails to make debugging easier"

This could be a good first issue.

Reply:
good idea i can go through some of these todos and make explicit issues?